Disable `enable_spectator_access` setting for limited plan realms.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit disables the setting in frontend for limited plan realms.
- Second commit adds code to set the setting to `False` on changing plan to limited.

Fixes: <!-- Issue link, or clear description.--> https://chat.zulip.org/#narrow/stream/9-issues/topic/Web-public.20streams

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2022-06-03 13-26-34](https://user-images.githubusercontent.com/35494118/171874230-1cdd80b6-88f8-4509-9bd6-67abc24f0c94.png)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on zerver/actions/create_realm.py:
```diff
@@ -161,6 +161,8 @@ def do_create_realm(
         kwargs["org_type"] = org_type
     if enable_spectator_access is not None:
+        if enable_spectator_access and plan_type == Realm.PLAN_TYPE_LIMITED:
+            enable_spectator_access = False
```
We pass `enable_spectator_access` and `plan_type` to `do_create_realm` only when calling it from `populate_db`. Also, for now I have kept it to set it as False in that case, rather than raising error.

timabbott replied:
Maybe better would be this?

`assert !enable_spectator_access or plan_type != Realm.PLAN_TYPE_LIMITED`

With a comment that we're enforcing the invariant that the LIMITED plan type isn't allow to have spectators enabled.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -442,12 +442,21 @@ def do_change_realm_plan_type(
+            "message_visibility_limit",
+            "upload_quota_gb",
+        ]
+    )
```
I think if we wanted to do that, the right mechanism would be to call `do_set_realm_property`; which might be a good way to implement this from an abstractions standpoint anyway. I think a coherent way to do it would be to call that from the top of `do_change_plan_type`, before we do the other changes, basically starting with:
```
if plan_type == Realm.PLAN_TYPE_LIMITED and realm.enable_spectator_access:
     # comment
     do_set_realm_property(...)
```

timabbott commented:
@alya thoughts on the visual design? It's a bit unfortunate that it's not 100% obvious from page structure that the upgrade tip only applies to the one disabled setting on the next line. I could imagine instead skipping the upgrade tip and just having that be a thing that we discuss clearly on the linked /help/ page.

timabbott commented on zerver/lib/events.py:
```diff
@@ -960,6 +960,8 @@ def _draft_update_action(i: int) -> None:
                 state["zulip_plan_is_not_limited"] = event["value"] != Realm.PLAN_TYPE_LIMITED
                 state["realm_upload_quota_mib"] = event["extra_data"]["upload_quota"]
+                if event["value"] == Realm.PLAN_TYPE_LIMITED:
+                    state["realm_enable_spectator_access"] = False
```
I think we'd get to cut this line if we do the `do_set_realm_property` approach noted above, which is one reason I kinda want to do that.

timabbott commented:
Thanks for working on this @sahil839! I posted a fairly detailed batch of comments.

alya commented:
> Left a couple of comments. Also a couple of points -
> 
> * We do not change the setting if we are disabling them for other settings which are not allowed on limited plans like message retention, etc.
> * Also, do we want to reset the setting value to the value that was before we disabled the setting when upgrading again from limited plan.

Sorry, can you explain in more detail what you mean about other settings?

I guess this is a bit tricky. I noticed that in the screenshot the checkbox is greyed out but checked. Would that be interpreted as the setting being "off", or as it being "on" and impossible to turn off?

Anyway, maybe a more detailed explanation of how message retention, etc. works would help. :)

alya commented:
> @alya thoughts on the visual design? It's a bit unfortunate that it's not 100% obvious from page structure that the upgrade tip only applies to the one disabled setting on the next line. I could imagine instead skipping the upgrade tip and just having that be a thing that we discuss clearly on the linked /help/ page.

Mm, I didn't feel like it was especially confusing when playing with the PR earlier. Maybe we could change the text to "Setting available on Zulip Cloud Standard..." to try to clarify the scope?

I think having the banner there is better than relegating the info to /help.

sahil839 commented:
Updated the PR -
- Call `do_set_realm_property` in `do_change_realm_plan_type` to change `enable_spectator_access`.
- Added assertion statements in `do_create_realm` to test both the cases where `plan_type` is explicitly passed as limited and when no value of `plan_type` is passed and we assume limited plan if `settings.BILLING_ENABLED` is `True`.
- Extended `test_fetch_raw_message_spectator` test to check the case where a limited plan realm has `enable_spectator_access` as `True`. Do we want to add code to the view function to not allow setting `enable_spectator_access` as `True` on limited realms?

Still holding off for the change to either change the banner text or something else to mention clearly that this banner is only for that setting.

sahil839 commented:
> > Left a couple of comments. Also a couple of points -
> > 
> > * We do not change the setting if we are disabling them for other settings which are not allowed on limited plans like message retention, etc.
> > * Also, do we want to reset the setting value to the value that was before we disabled the setting when upgrading again from limited plan.
> 
> Sorry, can you explain in more detail what you mean about other settings?
> 
> I guess this is a bit tricky. I noticed that in the screenshot the checkbox is greyed out but checked. Would that be interpreted as the setting being "off", or as it being "on" and impossible to turn off?
> 
> Anyway, maybe a more detailed explanation of how message retention, etc. works would help. :)

Ohh,the screenshot above was outdated, I have updated the screenshot, since we set the setting to `False` in server on downgrading the plan.

And regarding the message retention setting, suppose initally the realm is on `Standard` plan and realm message retention setting is set to some `x` number of days. Then if realm is downgraded to `Limited` plan, the setting is disabled in the UI but value still remain that `x` number of days and we do not change it again to the default value of forever like we are doing for spectator access setting.

alya commented:
Ah, got it, thanks!

Yeah, OK, it's not perfect, but I think the simplest approach is to err on the side of making sure an org doesn't accidentally expose any data. You wouldn't want to be surprised by web-public streams being turned on if your org stopped paying for Zulip Standard a year ago and you've forgotten all about this feature.

So I would say let's go with what I believe you have -- disabling the setting when the org is downgraded, and keeping it disabled when you upgrade again.

@timabbott does that seem reasonable?

timabbott commented:
> So I would say let's go with what I believe you have -- disabling the setting when the org is downgraded, and keeping it disabled when you upgrade again.

That sounds correct to me. 

> Do we want to add code to the view function to not allow setting enable_spectator_access as True on limited realms?

Yeah, that would probably be the most consistent thing to do. It can be a follow-up this week.

> And regarding the message retention setting, suppose initially the realm is on Standard plan and realm message retention setting is set to some x number of days. Then if realm is downgraded to Limited plan, the setting is disabled in the UI but value still remain that x number of days and we do not change it again to the default value of forever like we are doing for spectator access setting.

Because this setting is custom configuration for each stream as well as the realm as a whole, it'd be fairly annoying to restore whatever you'd configured when upgrading again. So I'd prefer to not **automatically** do something here.
